### PR TITLE
netcdf-mpi 4.7.4 (new formula)

### PR DESCRIPTION
**Pull request by burlen**
This adds a formula for NetCDF "parallel 4", i.e. NetCDF compiled with HDF5
parallel MPI I/O enabled.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by burlen** [comment]
can someone more familiar with this process advise on how to handle a file installed by the library?

the error reported is
```
==> FAILED
Error: 1 problem in 1 formula detected
netcdf-mpi:
  * Non-libraries were installed to "/usr/local/opt/netcdf-mpi/lib"
    Installing non-libraries to "lib" is discouraged.
    The offending files are:
      /usr/local/opt/netcdf-mpi/lib/libnetcdf.settings
```

**Comment by SMillerDev** [comment]
Why does it install that file there?

**Comment by burlen** [comment]
> Why does it install that file there?

I believe it is a convention with autoconf to document how the library was built.

In some other other formula the install location was changed by search and replace in the makefile itself. I have now done this and all seems OK.

**Comment by burlen** [comment]
I think this is ready for a review

File: Formula/netcdf-mpi.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  depends_on "hdf5-mpi"
+
+  conflicts_with "netcdf", :because => "netcdf-mpi is an MPI parallel build, use either netcdf or netcdf-mpi"
+
+  def install
+    # move files that NetCDF wants to place in libdir,
+    # because brew does not want them there.
+    inreplace "./Makefile.in",
+          "settingsdir = $(libdir)",
+          "settingsdir = #{pkgshare}"
```
Comment (SMillerDev) [inline Formula/netcdf-mpi.rb:17]: If it's settings it should be `#{etc}` I think.

File: Formula/netcdf-mpi.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    system "./configure", "--enable-shared",
+                          "--enable-static",
+                          "--disable-dap",
+                          "--enable-netcdf-4",
+                          "--enable-parallel4",
+                          "--prefix=#{prefix}",
+                          "CC=mpicc"
```
Comment (SMillerDev) [inline Formula/netcdf-mpi.rb:25]: I think you need to use `ENV.CC` for this.

File: Formula/netcdf-mpi.rb (hunk 1 above)
Comment (burlen) [reply Formula/netcdf-mpi.rb:17]: OK, will make this change but btw, I got this solution from the `hdf5-mpi` formula.
https://github.com/Homebrew/homebrew-core/blob/f2ba900dbe72b43943b981aebcc0f2eb1760cc25/Formula/hdf5-mpi.rb#L33
perhaps that needs to be fixed as well.

File: Formula/netcdf-mpi.rb (hunk 2 above)
Comment (burlen) [reply Formula/netcdf-mpi.rb:25]: OK, will make this change, but btw, I got this solution from the `hdf5-mpi` formula.
https://github.com/Homebrew/homebrew-core/blob/f2ba900dbe72b43943b981aebcc0f2eb1760cc25/Formula/hdf5-mpi.rb#L45
perhaps that needs to be fixed as well.

Comment (burlen) [reply Formula/netcdf-mpi.rb]: done.

File: Formula/netcdf-mpi.rb (hunk 1 above)
Comment (burlen) [reply Formula/netcdf-mpi.rb]: done.

File: Formula/netcdf-mpi.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
+class NetcdfMpi < Formula
+  desc "The NetCDF C I/O library compiled with MPI parallel features enabled"
+  homepage "https://www.unidata.ucar.edu/software/netcdf/"
+  url "https://www.unidata.ucar.edu/downloads/netcdf/ftp/netcdf-c-4.7.4.tar.gz"
+  sha256 "0e476f00aeed95af8771ff2727b7a15b2de353fb7bb3074a0d340b55c2bd4ea8"
+  license "BSD-3-Clause"
+
+  depends_on "hdf5-mpi"
+
+  conflicts_with "netcdf", :because => "netcdf-mpi is an MPI parallel build, use either netcdf or netcdf-mpi"
```
Comment (sjackman) [inline Formula/netcdf-mpi.rb:10]:
```suggestion
  keg_only "netcdf-mpi is an MPI parallel build of netcdf"
```
With `conflicts_with "netcdf"`, you won't be able to install anything that depends on `netcdf`. Instead, make this formula `keg_only`.

File: Formula/netcdf-mpi.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  def install
+    # use the C compiler wrapper
+    ENV["CC"] = "mpicc"
+
+    # move files that NetCDF wants to place in libdir,
+    # because brew does not want them there. etc was
+    # recommended during the code review
+    inreplace "./Makefile.in",
+          "settingsdir = $(libdir)",
+          "settingsdir = #{etc}/netcdf-mpi"
```
Comment (sjackman) [inline Formula/netcdf-mpi.rb:21]:
```suggestion
          "settingsdir = #{pkgshare}"
```
Are these files intended to be modified by the user? If not, they ought to go in `pkgshare`.

File: Formula/netcdf-mpi.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  test do
+    assert_equal "yes", shell_output("#{bin}/nc-config --has-parallel4").strip
+
+    (testpath/"test.c").write <<~EOS
+      #include <stdio.h>
+      #include "netcdf_meta.h"
+      int main()
+      {
+        printf(NC_VERSION);
```
Comment (sjackman) [inline Formula/netcdf-mpi.rb:41]:
```suggestion
        puts(NC_VERSION);
```

Comment (sjackman) [reply Formula/netcdf-mpi.rb:41]: Because `NC_VERSION` is a macro, this test doesn't use any symbols from the `netcdf` library. Please call a function from the library.

File: Formula/netcdf-mpi.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        return 0;
+      }
+    EOS
+    system ENV.cc, "test.c", "-L#{lib}", "-I#{include}", "-lnetcdf",
+                   "-o", "test"
```
Comment (sjackman) [inline Formula/netcdf-mpi.rb:46]:
```suggestion
    system "mpicc", "test.c", "-I#{include}", "-L#{lib}", "-lnetcdf", "-o", "test"
```

**Review by sjackman — CHANGES_REQUESTED** [review]

File: Formula/netcdf-mpi.rb (hunk 3 above)
Comment (burlen) [reply Formula/netcdf-mpi.rb:13]: `keg_only` is not what we want - this install should be linked into `/usr/local` so that it's easily found and used.

formula that depend on `netcdf`'s install of the netcdf-c library can be updated to depend on either `netcdf` OR `netcdf-mpi`. Those codes will compile, link and run without any modification.

File: Formula/netcdf-mpi.rb (hunk 4 above)
Comment (burlen) [reply Formula/netcdf-mpi.rb]: agree. not sure why @SMillerDev wanted me to change it to #{etc}.

File: Formula/netcdf-mpi.rb (hunk 5 above)
Comment (burlen) [reply Formula/netcdf-mpi.rb:44]: agree. I added 2 tests that use the install to write simple files in serial and parallel

File: Formula/netcdf-mpi.rb (hunk 6 above)
Comment (burlen) [reply Formula/netcdf-mpi.rb]: since this is not a parallel code, `mpicc` isn't needed. part of netcdf design is that serial codes can compile and link with parallel library without modification.

**Comment by burlen** [comment]
OK, ready for review. I've incorporated @sjackman feed back.

File: Formula/netcdf-mpi.rb (hunk 3 above)
Comment (sjackman) [reply Formula/netcdf-mpi.rb:13]: Homebrew unfortunately has no mechanism to support alternate dependencies. There is no way to depend on `netcdf` OR `netcdf-mpi`.

File: Formula/netcdf-mpi.rb (hunk 4 above)
Comment (sjackman) [reply Formula/netcdf-mpi.rb]: Are these files intended to be modified by the user? Sean probably assumed they were. I suspect they're not.

**Comment by sjackman** [comment]
This PR for `netcdf-mpi` cannot be merged so long as it conflicts with `netcdf`, because someone with `netcdf-mpi` installed will not be able to install anything that depends `netcdf` so long as it conflicts with `netcdf`.

The options are…

1. Modify this `netcdf-mpi` to only install the files that do not conflict with `netcdf`, and make `netcdf-mpi` depend on `netcdf`, if this is technically possible. It may not be.
2. Make `netcdf-mpi` keg only. You may `brew install netcdf netcdf-mpi && brew unlink netcdf && brew link --force netcdf-mpi`
3. Create a third-party tap with the formula `netcdf-mpi` that conflicts with `netcdf`

**Comment by burlen** [comment]
> netcdf-mpi cannot be merged so long as it conflicts with netcdf

This seems like a fairly arbitrary point of view. Other formula do exactly this and it's not a problem. See `hdf5-mpi` and and `hdf5`.

> someone with netcdf-mpi installed will not be able to install anything that depends netcdf so long as netcdf-mpi conflicts with netcdf.

So what? If need parallel I/O capabilities I want `netcdf-mpi` and if I don't then `netcdf` is fine. I will choose the one that best satisfies my needs. I don't need you to tell me what formula I want or need.

Codes that make use of serial NetCDF C library can compile and link against a parallel NetCDF C library without modification. They could be installed with either `netcdf` or `netcdf-mpi. The fact that brew doesn't support this(assuming you are right in that) seems like a substantial issue in itself, one far beyond the reasonable scope of this pr

File: Formula/netcdf-mpi.rb (hunk 2 above)
Comment (sjackman) [reply Formula/netcdf-mpi.rb:29]: `./configure CC=mpicc` should be fine.

File: Formula/netcdf-mpi.rb (hunk 1 above)
Comment (sjackman) [reply Formula/netcdf-mpi.rb:20]: If these settings are meant to modified by the user then they should go in `etc`. If they are _not_ meant to be modified by the user, they should go in `pkgshare`.

**Comment by sjackman** [comment]
>> netcdf-mpi cannot be merged so long as it conflicts with netcdf

> This seems like a fairly arbitrary point of view. Other formula do exactly this and it's not a problem. See hdf5-mpi and and hdf5.

`hdf5-mpi 1.12.0 (new formula)` https://github.com/Homebrew/homebrew-core/pull/53606 by @ax3l merged by @chenrui333

Homebrew is maintained by a team of maintainers. Maintaining Homebrew is full of judgement calls, and we may not always agree, which is fine.

>> someone with netcdf-mpi installed will not be able to install anything that depends netcdf so long as netcdf-mpi conflicts with netcdf.

> So what? If need parallel I/O capabilities I want netcdf-mpi and if I don't then netcdf is fine. I will choose the one that best satisfies my needs. I don't need you to tell me what formula I want or need.

Please mind your tone. The maintainers of Homebrew are volunteers that do this work unpaid. We do our best to triage PRs and make choices that we feel best support the majority of Homebrew users.

If a user installs the `netcdf-mpi` formula from this PR, they will not then be able to install any of the 13 formulae that depend on `netcdf`, which I find a less than ideal situation.

> The fact that brew doesn't support this(assuming you are right in that) seems like a substantial issue in itself, one far beyond the reasonable scope of this pr

Perhaps, but Homebrew does not currently have this feature that you desire.

I gave you three possible ways forward in https://github.com/Homebrew/homebrew-core/pull/57732#issuecomment-661998212
I personally think that making `netcdf-mpi` keg only, and then force linking it if you so desire is a good path forward. Does that work for you?
